docs/dev_guide/frequently-asked-questions-faq.rst

```diff
@@ -162,6 +163,152 @@ expected results. Instead, exclude CUB library implementation source code from
+         if (sdata[tid] < sdata[tid + 4]) {
+            ind[tid] = ind[tid + 4];
+            sdata[tid] = sdata[tid + 4];
+         }
+         item_ct1.barrier(sycl::access::fence_space::local_space);
+         if (sdata[tid] < sdata[tid + 2]) {
+            ind[tid] = ind[tid + 2];
+            sdata[tid] = sdata[tid + 2];
+         }
+         item_ct1.barrier(sycl::access::fence_space::local_space);
+         if (sdata[tid] < sdata[tid + 1]) {
+            ind[tid] = ind[tid + 1];
+            sdata[tid] = sdata[tid + 1];
+         }
+         item_ct1.barrier(sycl::access::fence_space::local_space);
+      }
+   }
+
+is adjusted to move the synchronization statement outside of the 
+conditional statement:
+
+.. code-block:: cpp
+   :linenos:
+
+   // fixed SYCL code
+
+   void kernel(float *sdata, int *ind, const sycl::nd_item<3> &item_ct1) {
+      unsigned int tid = item_ct1.get_local_id(2);
+      if (tid < 32) {
+         if (sdata[tid] < sdata[tid + 32]) {
+            ind[tid] = ind[tid + 32];
+            sdata[tid] = sdata[tid + 32];
+         }
+      }
+      item_ct1.barrier(sycl::access::fence_space::local_space);
+      if (tid < 32) {
+         if (sdata[tid] < sdata[tid + 16]) {
+            ind[tid] = ind[tid + 16];
+            sdata[tid] = sdata[tid + 16];
+         }
+      }
+      item_ct1.barrier(sycl::access::fence_space::local_space);
+      if (tid < 32) {
+         if (sdata[tid] < sdata[tid + 8]) {
+            ind[tid] = ind[tid + 8];
+            sdata[tid] = sdata[tid + 8];
+         }
+      }
+      item_ct1.barrier(sycl::access::fence_space::local_space);
+      if (tid < 32) {
+         if (sdata[tid] < sdata[tid + 4]) {
+            ind[tid] = ind[tid + 4];
+            sdata[tid] = sdata[tid + 4];
+         }
+      }
+      item_ct1.barrier(sycl::access::fence_space::local_space);
+      if (tid < 32) {
+         if (sdata[tid] < sdata[tid + 2]) {
+            ind[tid] = ind[tid + 2];
+            sdata[tid] = sdata[tid + 2];
+         }
+      }
+      item_ct1.barrier(sycl::access::fence_space::local_space);
+      if (tid < 32) {
+         if (sdata[tid] < sdata[tid + 1]) {
+            ind[tid] = ind[tid + 1];
+            sdata[tid] = sdata[tid + 1];
+         }
+      }
+      item_ct1.barrier(sycl::access::fence_space::local_space);
+      }
+   }
+
+Barrier is used in loop statements. The following original code:
+
+.. code-block:: cpp
+   :linenos:
+
+   // original code
+
+   void compute(int np, int nd, const sycl::nd_item<3> &item_ct1) {
+      unsigned int curr_particle =
+         item_ct1.get_group(2) * item_ct1.get_local_range(2) +
+         item_ct1.get_local_id(2);
+      for (; curr_particle < np; curr_particle +=
+         item_ct1.get_group_range(2) * item_ct1.get_local_range(2)) {
+         ...
+         item_ct1.barrier();
+         ...
+      }
+   }
+
+is adjusted to fix the conditional in loop statement, let all thread

```

Comment:
The following is the manually-adjusted code of moving the synchronization statement out of the for loop.